# Individual Heterozygosity + Inbreeding 

```bash
cat bams_good | awk -F '/' '{print $2}' | awk -F '.' '{print $1}' | while read sample ;
do echo ${sample} ;
sbatch Hetero.sh ${sample} ;
done

```

-> Hetero.sh

```bash
#!/bin/bash

Bamfile=$(ls ../../*.bam | grep ${1})

module purge 
module load SAMtools/1.9-GCC-8.2.0-2.31.1
samtools faidx Assembly_Puffin_NU.MT.fasta

module load angsd/0.931-GCC-8.2.0-2.31.1

Threads="1"
GENOME_REF=Assembly_Puffin_NU.MT.fasta

Chromosome=$(cat ChromosomeList.edit | awk -F ':' '{print $1}' | grep -w Scaffolds_chromosome_${SLURM_ARRAY_TASK_ID} )
ChromosomeEdit=$(echo ${Chromosome}":")

grep -w Scaffolds_chromosome_${SLURM_ARRAY_TASK_ID} sites2do > sites_Chromosome${SLURM_ARRAY_TASK_ID}_2do
#sites2do comes from previous SFS estimation

sleep 120

angsd sites index sites_Chromosome${SLURM_ARRAY_TASK_ID}_2do

sleep 120 

#run angsd
angsd -i ${Bamfile} -anc ${GENOME_REF} -ref ${GENOME_REF} -doSaf 1 -GL 1 -r ${ChromosomeEdit} -sites sites_Chromosome${SLURM_ARRAY_TASK_ID}_2do -doCounts 1 -uniqueOnly 1 -remove_bads 1 -C 50 -baq 2 -minMapQ 30 -minQ 30 -P ${Threads} -out ${1}_Chromosome${SLURM_ARRAY_TASK_ID}

```

Estimation of individual SFS

```bash
cat ../bams_good | awk -F '/' '{print $2}' | awk -F '.' '{print $1}' | while read sample ;
do echo ${sample} ;
sbatch SFS.creation.sh ${sample} ;
done

```

-> SFS.creation.sh

```bash
#!/bin/bash

module load angsd/0.931-GCC-8.2.0-2.31.1
Threads="16"

realSFS cat ${1}_Chromosome*.saf.idx -outnames ${1} ;
realSFS ${1}.saf.idx -P ${Threads} -fold 1 > ${1}.sfs ;

```

Heterozygosity Calculation

```bash

#in R
module purge
module load R/3.6.0-intel-2019a

touch Heterozygosity.txt
cat ../bams_good | awk -F '/' '{print $2}' | awk -F '.' '{print $1}' | while read sample ;
do echo ${sample} ;
Hetero=$(Rscript Hetero.R ${sample}.sfs) ;
echo -e ${sample}"\t"${Hetero} >> Heterozygosity.txt ;
done ;

```

-> Hetero.R

```r
infile <- commandArgs(trailingOnly = TRUE)
a<-scan(infile)
hetero <- a[2]/sum(a)
cat(hetero)
```



Inbreeding estimation based on RoH

```bash

###### make Windows file

module purge
module load angsd/0.931-GCC-8.2.0-2.31.1
Threads="16"
sample=BJO001
realSFS ${sample}.saf.idx ${sample}.saf.idx -P ${Threads} -fold 1 > test.ml
realSFS fst index ${sample}.saf.idx ${sample}.saf.idx -sfs test.ml -whichFst 1 -fstout test -fold 1
realSFS fst stats2 test.fst.idx -win 100000 -step 50000 -fold 1 > slidingwindow #this will determine RoH size
rm test.fst.gz
touch Regions.txt
cat slidingwindow | tail -n+2 | awk '{print $2}' | uniq | while read chromo ;
do echo ${chromo} ;
echo -e ${chromo}":1-100000" >> Regions.txt ;
grep -w ${chromo} slidingwindow | while read line ;
do Region=$(echo ${line} | awk '{print $1}' | awk -F '(' '{print $4}' | sed 's/)//g' | sed 's/,/-/g') ;
echo -e ${chromo}":"${Region} >> Regions.txt ;
done ;
done ;

######


cat bams_good | awk -F '/' '{print $2}' | awk -F '.' '{print $1}' | while read sample ;
do echo ${sample} ;
sbatch realSFS_ROH.sh ${sample} ;
done ;

```

-> realSFS_ROH.sh

```bash
#!/bin/bash

module load angsd/0.931-GCC-8.2.0-2.31.1

Threads="1"

touch ${1}_${SLURM_ARRAY_TASK_ID}.sfs

NumberOfLines=$(echo "742") 
BottomLine=$(bc <<<"scale=0; ${SLURM_ARRAY_TASK_ID} * 742 ") 

cat Regions.txt | head -${BottomLine} | tail -${NumberOfLines} | while read line ;
do realSFS ${1}.saf.idx -r ${line} -P ${Threads} -fold 1 >> ${1}_${SLURM_ARRAY_TASK_ID}.sfs ;
done ;

```

Combine the individual window-based sfs's

```bash
cat bams_good | awk -F '/' '{print $2}' | awk -F '.' '{print $1}' | while read sample ;
do echo ${sample} ;
touch ${sample}.sfs ;
for i in {1..31} ;
do cat ${sample}_${i}.sfs >> ${sample}.sfs ;
rm ${sample}_${i}.sfs ;
done ;
done ;

```
Calculate heterozygosity per window
->HeteroRoh.R

```bash
#####HeteroRoh.R######
infile <- commandArgs(trailingOnly = TRUE)
sample <- substr(infile, 1, 6)
df <- as.data.frame(read.table(file = infile, stringsAsFactors = FALSE))
df$V4 <- rowSums(df)
df$V5 <- df$V2 / df$V4
df$V6 <- sample
sample_table <- as.data.frame(cbind(df$V6, df$V5))
write.table(sample_table, file = paste(sample,"heterozygosity.txt", sep = "_"), col.names = F, row.names = F, quote = F, sep = "\t")
###############

module purge
module load R/3.6.0-intel-2019a
cat bams_good | awk -F '/' '{print $2}' | awk -F '.' '{print $1}' | while read sample ;
do echo ${sample} ;
paste Regions.txt ${sample}.sfs > ${sample}.regionSFS ;
Rscript HeteroRoh.R ${sample}.sfs ;
paste ${sample}.regionSFS ${sample}_heterozygosity.txt > ${sample}.RoH
done ;

cat *heterozygosity.txt > HeterozygosityCombined.txt

#choose homozygosity cutoff by plotting

```

Assess RoH lengths and F_RoH
-> ROH_Length_Count.R

```bash
##### ROH_Length_Count.R ####
library(tidyr)
library(tidyverse)
infile <- commandArgs(trailingOnly = TRUE)
sample <- substr(infile, 1, 6)
df <- as.data.frame(read.table(file = infile, stringsAsFactors = FALSE))
df_edit <- cbind(df$V5, data.frame(do.call('rbind', strsplit(as.character(df$V1),':',fixed=TRUE))), df$V6)
colnames(df_edit) <- c("Sample", "Chromosome", "Region", "Hetero")
for (i in 1:25) {
df_edit2 <- df_edit %>% filter(Chromosome == paste("Scaffolds_chromosome",i, sep = "_"))
df_edit2$ROH <- ifelse(df_edit2$Hetero < 0.001435663, 1, 0) # enter cutoff here
value <- rle(df_edit2$ROH)$values
lengths <- as.numeric(rle(df_edit2$ROH)$lengths)
df2 <- cbind.data.frame(value, lengths)
df2$chromosome <- paste("Chromosome",i, sep = "_")
df3 <- df2 %>% filter(value == 1) %>% filter(lengths > 1)
df3$ROH_Length <- df3$lengths * 0.1 - ((df3$lengths - 1) * 0.05 )
assign(paste("Chromo",i,"df3", sep = "_"), df3)
}
Dataframes <- lapply(paste("Chromo_",1:25,"_df3", sep = ""), get)
All_chromoMerge_df3 <- bind_rows(Dataframes)
All_chromoMerge_df3$value <- sample
colnames(All_chromoMerge_df3) <- c("Sample", "Number", "Chromosome","ROH_Length")
write.table(All_chromoMerge_df3, file = paste(sample,"_ROH_Lengths.txt", sep = ""), row.names = F, quote = F, sep = "\t")
sample_inbreeding <- unname(colSums(All_chromoMerge_df3[,"ROH_Length", drop = FALSE]) / (nrow(df_edit) * 0.1 - ((nrow(df_edit) - 1) * 0.05)))
Inbreeding_table <- as.data.frame(cbind(sample, sample_inbreeding))
write.table(Inbreeding_table, file = paste(sample,"_F_ROH.txt", sep = ""), row.names = F, quote = F, sep = "\t")
###############

module purge
module load R/3.6.0-intel-2019a

echo -e "Sample\tNumber\tChromosome\tROH_Length" > AllSamples_ROH_Length
echo -e "Sample\tF_ROH" > AllSamples_F_ROH
cat bams_good | awk -F '/' '{print $2}' | awk -F '.' '{print $1}' | while read sample ;
do echo ${sample} ;
Rscript ROH_Length_Count.R ${sample}.RoH ;
cat ${sample}_ROH_Lengths.txt | tail -n+2 >> AllSamples_ROH_Length ;
cat ${sample}_F_ROH.txt | tail -n+2 >> AllSamples_F_ROH ;
done ;

```

Plotting

```bash
library(ggplot2)
library(tidyr)
library(tidyverse)
library(ggpubr)
library(rstatix)
library(dplyr)


colorcodes=c("#cab2d6", "#fb6a4a", "#99000d", "#9f9f9f", "#fc9272", "#606060",
              "#fedccbff", "#ffaa00", "#ef3b2c", "#fcbba1", "#1f78b4", "#cb181d")



colonies <- c("Bjørnøya", "Breiðafjörður", "Faroe", "Gannet Isl.", "Grímsey", 
              "Gull Isl.", "Hornøya" , "Isle of May", "Papey" ,
              "Røst", "Spitsbergen", "Vestmannaeyjar")

## Plot sample-based heterozygosity
df <- as.data.frame(read.table(file = "HeterozygosityCombined.txt", stringsAsFactors = FALSE))
df$V3 <- substr(df$V1, 1, 3)
quantile(df$V2, probs = seq(0, 1, 1/10))

y <- c("SPI", "BJO", "HOR", "ROS", "WES", "GRI", "BRE", "PAP", "FAR", "IOM", "GAN", "GUL")
df_edit <- df[order(match(df$V3,y)),]
rownames(df_edit)<-c(1:nrow(df_edit))

df_edit$V1 <- factor(df_edit$V1, levels = unique(df_edit$V1),ordered = TRUE)

ggplot(df_edit, aes(x=V1, y=V2, fill=V3)) + 
  geom_boxplot(outlier.shape = NA, aes(alpha = 1)) +
  stat_boxplot(geom = 'errorbar') +
  theme_bw() +
  theme(axis.title = element_text(size = 16, face = "bold"), legend.position = "none",
        panel.grid = element_blank(), axis.text.x = element_text(size = 8, angle = 60, hjust = 1)) +
  scale_fill_manual(values = colorcodes) +
  xlab("Samples") +
  ylab("Heterozygosity") +
  geom_hline(yintercept = 0.001435663, linetype="dashed", color = "red") +
  ylim(0,0.005)
  
df_2 <- as.data.frame(read.table(file = "AllSamples_ROH_Length", stringsAsFactors = FALSE, header = TRUE))
df_2$pop <- substring(df_2$Sample, 1, 3)
df_2$pop <- as.factor(df_2$pop) #make pops a factor
names(colorcodes) <- levels(df_2$pop) #assign colors to colonies
names(colonies) <- levels(df_2$pop) #assign colony names to colonies
df_2$pop <- factor(df_2$pop, levels=c('SPI', 'BJO', 'HOR', 'ROS', 'WES', 'GRI', 'BRE', 'PAP', 'FAR', 'IOM', 'GAN', 'GUL')) #sort after specific order
 
df_3 <- as.data.frame(read.table(file = "AllSamples_F_ROH", stringsAsFactors = FALSE, header = TRUE))
df_3$pop <- substring(df_3$Sample, 1, 3)
df_3$pop <- as.factor(df_3$pop) #make pops a factor
names(colorcodes) <- levels(df_3$pop) #assign colors to colonies
names(colonies) <- levels(df_3$pop) #assign colony names to colonies
df_3$pop <- factor(df_3$pop, levels=c('SPI', 'BJO', 'HOR', 'ROS', 'WES', 'GRI', 'BRE', 'PAP', 'FAR', 'IOM', 'GAN', 'GUL')) #sort after specific order
stat.test_edit <- dunn_test(df_3, F_ROH ~ pop, p.adjust.method ="holm", detailed = TRUE) %>% filter(p.adj.signif == "*" | p.adj.signif == "**" | p.adj.signif == "***" | p.adj.signif == "****") 


#PANEL 
table_3 <- read.table(file = "Heterozygosity.txt", sep = "\t", header = FALSE)
mat_3 <- as.data.frame(table_3)
mat_3$pop <- substring(mat_3$V1, 0, 3)
mat_3$pop <- as.factor(mat_3$pop) #make pops a factor
names(colorcodes) <- levels(mat_3$pop) #assign colors to colonies
names(colonies) <- levels(mat_3$pop) #assign colony names to colonies
mat_3$pop <- factor(mat_3$pop, levels=c('SPI', 'BJO', 'HOR', 'ROS', 'WES', 'GRI', 'BRE', 'PAP', 'FAR', 'IOM', 'GAN', 'GUL')) #sort after specific order
stat.test_edit.2 <- dunn_test(mat_3, V2 ~ pop, p.adjust.method ="holm", detailed = TRUE) %>% filter(p.adj.signif == "*" | p.adj.signif == "**" | p.adj.signif == "***" | p.adj.signif == "****") 


image1 <- ggplot(mat_3, aes(x=pop, y=V2)) +
  geom_boxplot(outlier.shape = NA, aes(fill=pop)) +
  theme_bw() +
  theme(axis.title = element_text(size = 16, face = "bold"), legend.position = "none",
        panel.grid = element_blank()) +
  scale_fill_manual(values = colorcodes) +
  scale_x_discrete(labels = colonies) +
  xlab("Colonies") +
  ylab("Heterozygosity") +
  geom_jitter(shape=16, position=position_jitter(0.1), size = 1, color = "black") +
  stat_compare_means(label.x = 6, label.y = 0.00225)# +
  #stat_pvalue_manual(stat.test_edit.2, label = "p.adj.signif", remove.bracket = FALSE, 
                     y.position = c(0.00255, 0.002575, 0.0026, 0.002625, 0.00265, 0.002675, 0.0027, 0.002575), 
                     size = 6, color = "red", tip.length = 0.005)

image2 <- ggplot(df_3, aes(x=pop, y=F_ROH)) + 
  geom_boxplot(outlier.shape = NA, aes(fill=pop)) +
  theme_bw() +
  theme(axis.title = element_text(size = 14, face = "bold"), legend.position = "none",
        panel.grid = element_blank(), axis.text.x = element_text(size = 10, angle = 45, hjust = 1)) +
  scale_fill_manual(values = colorcodes) +
  scale_x_discrete(labels = colonies) +
  xlab("Colonies") +
  ylab("F(RoH)") +
  geom_jitter(shape=16, position=position_jitter(0.1), size = 1, color = "black") +
  stat_compare_means(label.x = 6, label.y = 0.16)# + #Global Kruskal
  stat_pvalue_manual(stat.test_edit, label = "p.adj.signif", remove.bracket = FALSE, 
                     y.position = c(0.185, 0.19, 0.195, 0.2, 0.205, 0.125, 0.14, 0.145, 0.15), 
                     size = 6, color = "red", tip.length = 0.005)

image3 <- ggplot(df_2, aes(x=pop, y=ROH_Length, fill=pop)) + 
  geom_jitter(shape=16, position=position_jitter(0.1), size = 0.5, color = "black", alpha = 0.6) +
  geom_violin(scale = "count", adjust = 0.5) +
  theme_bw() +
  theme(axis.title = element_text(size = 14, face = "bold"), legend.position = "none",
        panel.grid = element_blank(), axis.text.x = element_text(size = 10, angle = 45, hjust = 1)) +
  scale_fill_manual(values = colorcodes) +
  scale_x_discrete(labels = colonies) +
  xlab("Colonies") +
  ylab("RoH Length (Mbp)\nOnly RoHs > 500 kbp") +
  coord_cartesian(ylim = c(0.5,5.8)) +
  ylim(0.5, 10) #+
  #annotate("text", label = "Only RoHs > 500 kbp", x=5, y=4)

pdf("Puffin_Heterozygosit_Inbreeding_RoH_edit.pdf", 11, 15)
ggarrange(image1 + rremove("x.text") + rremove("x.title") + rremove("x.ticks"), 
          image2 + rremove("x.text") + rremove("x.title") + rremove("x.ticks"), 
          image3 + rremove("x.title"), 
          ncol = 1, nrow = 3, heights = c(1, 1, 1.3), align = "v")
dev.off()
```