# PCA, SNP weight/Selection



```bash
module purge
module load PCAngsd/200115-foss-2019a-Python-2.7.15

pcangsd.py -beagle ModernPuffinAngsd_NoZ_NoUnpl.beagle.ldpruned.gz -snp_weights -selection -sites_save -o ModernPuffin.NoZNoUnpl.pruned.pcangsd -threads 16 > PCAngsd.log

#do the same with subsets
mkdir FineScale
cd FineScale
cp ../ModernPuffinAngsd_NoZ_NoUnpl.beagle.ldpruned.gz .

#Need to change the header!
cat bams_good | awk -F '.' '{print $1}' > inds #individual IDs
gunzip ModernPuffinAngsd_NoZ_NoUnpl.beagle.ldpruned.gz
head -1 ModernPuffinAngsd_NoZ_NoUnpl.beagle.ldpruned > new.header
cat inds | while read line ;
do LineNumber=$(grep -n ${line} inds | cut -f1 -d:)
IndNumber=$(expr ${LineNumber} - 1)
sed -i "s/\bInd${IndNumber}\b/${line}/g" new.header ;
done 
tail -n+2 ModernPuffinAngsd_NoZ_NoUnpl.beagle.ldpruned | cat new.header - > ModernPuffinAngsd_NoZ_NoUnpl.beagle.ldpruned.edit

#which sites
awk -F '\t' '{print $1"\t"$2"\t"$3}' ModernPuffinAngsd_NoZ_NoUnpl.beagle.ldpruned.edit > ModernPuffinAngsd_NoZ_NoUnpl.beagle.ldpruned.edit.sites

# make subset1 beagle file
for pop in BRE FAR GRI HOR IOM PAP ROS WES GAN GUL ;
do echo ${pop} ;
grep ${pop} inds > ${pop}.inds ;
cat ${pop}.inds | while read line ;
do head -1 ModernPuffinAngsd_NoZ_NoUnpl.beagle.ldpruned.edit | tr '\t' '\n' | cat -n | grep ${line} | awk '{print $1}' | sed 's/ //g' > columnnumber.${line} ;
cat columnnumber.${line} | while read number ;
do cut -f ${number} ModernPuffinAngsd_NoZ_NoUnpl.beagle.ldpruned.edit > ModernPuffinAngsd_NoZ_NoUnpl.beagle.ldpruned.edit.subset1.number.${number} ;
done ;
rm columnnumber.${line} ;
paste ModernPuffinAngsd_NoZ_NoUnpl.beagle.ldpruned.edit.subset1.number.* > ModernPuffinAngsd_NoZ_NoUnpl.beagle.ldpruned.edit.subset1.ind.${line} ;
rm ModernPuffinAngsd_NoZ_NoUnpl.beagle.ldpruned.edit.subset1.number* ;
done ;
paste ModernPuffinAngsd_NoZ_NoUnpl.beagle.ldpruned.edit.subset1.ind.* > ModernPuffinAngsd_NoZ_NoUnpl.beagle.ldpruned.edit.subset1.pop.${pop} ;
rm ModernPuffinAngsd_NoZ_NoUnpl.beagle.ldpruned.edit.subset1.ind.* ;
done ;
paste ModernPuffinAngsd_NoZ_NoUnpl.beagle.ldpruned.edit.sites ModernPuffinAngsd_NoZ_NoUnpl.beagle.ldpruned.edit.subset1.pop.* > ModernPuffinAngsd_NoZ_NoUnpl.beagle.ldpruned.edit.subset1
cat *.inds > inds.subset1
rm *.inds
rm ModernPuffinAngsd_NoZ_NoUnpl.beagle.ldpruned.edit.subset1.pop*
gzip ModernPuffinAngsd_NoZ_NoUnpl.beagle.ldpruned.edit.subset1

# make subset2 beagle file

for pop in BRE FAR GRI HOR PAP ROS WES ;
do echo ${pop} ;
grep ${pop} inds > ${pop}.inds ;
cat ${pop}.inds | while read line ;
do head -1 ModernPuffinAngsd_NoZ_NoUnpl.beagle.ldpruned.edit | tr '\t' '\n' | cat -n | grep ${line} | awk '{print $1}' | sed 's/ //g' > columnnumber.${line} ;
cat columnnumber.${line} | while read number ;
do cut -f ${number} ModernPuffinAngsd_NoZ_NoUnpl.beagle.ldpruned.edit > ModernPuffinAngsd_NoZ_NoUnpl.beagle.ldpruned.edit.subset2.number.${number} ;
done ;
rm columnnumber.${line} ;
paste ModernPuffinAngsd_NoZ_NoUnpl.beagle.ldpruned.edit.subset2.number.* > ModernPuffinAngsd_NoZ_NoUnpl.beagle.ldpruned.edit.subset2.ind.${line} ;
rm ModernPuffinAngsd_NoZ_NoUnpl.beagle.ldpruned.edit.subset2.number* ;
done ;
paste ModernPuffinAngsd_NoZ_NoUnpl.beagle.ldpruned.edit.subset2.ind.* > ModernPuffinAngsd_NoZ_NoUnpl.beagle.ldpruned.edit.subset2.pop.${pop} ;
rm ModernPuffinAngsd_NoZ_NoUnpl.beagle.ldpruned.edit.subset2.ind.* ;
done ;
paste ModernPuffinAngsd_NoZ_NoUnpl.beagle.ldpruned.edit.sites ModernPuffinAngsd_NoZ_NoUnpl.beagle.ldpruned.edit.subset2.pop.* > ModernPuffinAngsd_NoZ_NoUnpl.beagle.ldpruned.edit.subset2
cat *.inds > inds.subset2
rm *.inds
rm ModernPuffinAngsd_NoZ_NoUnpl.beagle.ldpruned.edit.subset2.pop*
gzip ModernPuffinAngsd_NoZ_NoUnpl.beagle.ldpruned.edit.subset2

module purge
module load PCAngsd/200115-foss-2019a-Python-2.7.15

pcangsd.py -beagle ModernPuffinAngsd_NoZ_NoUnpl.beagle.ldpruned.edit.subset1.gz -snp_weights -selection -sites_save -o ModernPuffin.NoZNoUnpl.pruned.subset1.pcangsd -threads 16 > PCAngsd.subset1.log
pcangsd.py -beagle ModernPuffinAngsd_NoZ_NoUnpl.beagle.ldpruned.edit.subset2.gz -snp_weights -selection -sites_save -o ModernPuffin.NoZNoUnpl.pruned.subset2.pcangsd -threads 16 > PCAngsd.subset2.log

```

Plotting (1 example in R)

```bash
library(ggplot2)
library(ggrepel)
library(RcppCNPy)
library(tidyr)

bams=read.table("bams_good")[,1] # list of bam files
bams=sub(".Assembly_Puffin_NU.sorted.bam","",bams) #gets rid of unneccessary parts of filenames
colonies=substr(bams, 1, 3)
colorcodes=c("#a6cee3", "#d9d9d9", "#b2df8a", "#33a02c", "#fb9a99", "#fdbf6f", 
             "#e31a1c", "#ff7f00", "#cab2d6", "#6a3d9a", "#1f78b4", "#b15928")

###read in covariance matrix + selection + position file###
S <- npyLoad("ModernPuffin.NoThu.NoZNoUnpl.pruned.pcangsd.selection.npy")
C <- as.matrix(read.table("ModernPuffin.NoThu.NoZNoUnpl.pruned.pcangsd.cov"))
p<-read.table("ModernPuffin.NoThu.NoZNoUnpl.pruned.pcangsd.sites.edit",colC=c("factor","integer"),sep="_")
w<-npyLoad("ModernPuffin.NoThu.NoZNoUnpl.pruned.pcangsd.weights.npy")
write.table(w, file = "ModernPuffin.NoThu.NoZNoUnpl.pruned.pcangsd.weights.txt")
e <- eigen(C)

e_df <- as.data.frame(e$vectors, row.names = NULL)
e_d <- cbind(bams, data.frame(e_df, row.names=NULL))
pca <- cbind(e_d, colonies)

e.var.per <- round(e$values/sum(e$values)*100, 2)
barplot(e.var.per)

ggplot(data=pca, aes(x=V1, y=V2, fill=colonies, label=bams)) +
  #geom_label_repel(show.legend = FALSE)+
  geom_point(shape = 21, colour= "black", size=3 ) +
  xlab(paste("PC1 ( ", e.var.per[1], "% of observed genetic variation )", sep="")) +
  ylab(paste("PC2 ( ", e.var.per[2], "% of observed genetic variation )", sep="")) +
  theme_bw() +
  theme(panel.background = element_blank(),panel.border=element_rect(fill=NA),panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),strip.background=element_blank(),axis.text.x=element_text(colour="black"),
        axis.text.y=element_text(colour="black"),axis.ticks=element_line(colour="black"),
        plot.margin=unit(c(1,1,1,1),"line"), plot.title = element_text(hjust = 0.5, size=16, face="bold"), 
        axis.title = element_text(size=12, face="bold")) +
  labs(fill = "Colonies")  +
  scale_fill_manual(values = colorcodes, labels = c("Bjørnøya", "Breiðafjörður", "Faroer", "Gannet Isl.", "Grímsey", "Gull Isl.", "Hornøya" , "Isle of May", "Papey" ,
                                                     "Røst", "Spitsbergen", "Vestmannaeyjar")) +
  ggtitle("PCA (from Cov. Matrix - PCAngsd)")

#############################
#SELECTION ANALYSES - based on evomics http://evomics.org/learning/population-and-speciation-genomics/2018-population-and-speciation-genomics/angsd-activity-pca-mds/#sec9
#############################

## function for QQplot
qqchi<-function(x,...){
  lambda<-round(median(x)/qchisq(0.5,1),2)
  qqplot(qchisq((1:length(x)-0.5)/(length(x)),1),x,ylab="Observed",xlab="Expected",...);abline(0,1,col=2,lwd=2)
  legend("topleft",paste("lambda=",lambda))
}

### vectorize seleciton statistics (chi2 distributed)
s1<-as.vector(S[,1])
s2<-as.vector(S[,2])
## make QQ plot to QC the test statistics, takes 5 min
qqchi(s1)
qqchi(s2)

# convert test statistic to p-value
pval1<-1-pchisq(s1,1)
pval2<-1-pchisq(s2,1)

### MAKE pvalue QQ plot ###
# based on https://genome.sph.umich.edu/wiki/Code_Sample:_Generating_QQ_Plots_in_R
library(lattice)
qqunif.plot<-function(pvalues, 
                      should.thin=T, thin.obs.places=2, thin.exp.places=2, 
                      xlab=expression(paste("Expected (",-log[10], " p-value)")),
                      ylab=expression(paste("Observed (",-log[10], " p-value)")), 
                      draw.conf=TRUE, conf.points=1000, conf.col="lightgray", conf.alpha=.05,
                      already.transformed=FALSE, pch=20, aspect="iso", prepanel=prepanel.qqunif,
                      par.settings=list(superpose.symbol=list(pch=pch)), ...) {
  
  
  #error checking
  if (length(pvalues)==0) stop("pvalue vector is empty, can't draw plot")
  if(!(class(pvalues)=="numeric" || 
       (class(pvalues)=="list" && all(sapply(pvalues, class)=="numeric"))))
    stop("pvalue vector is not numeric, can't draw plot")
  if (any(is.na(unlist(pvalues)))) stop("pvalue vector contains NA values, can't draw plot")
  if (already.transformed==FALSE) {
    if (any(unlist(pvalues)==0)) stop("pvalue vector contains zeros, can't draw plot")
  } else {
    if (any(unlist(pvalues)<0)) stop("-log10 pvalue vector contains negative values, can't draw plot")
  }
  
  
  grp<-NULL
  n<-1
  exp.x<-c()
  if(is.list(pvalues)) {
    nn<-sapply(pvalues, length)
    rs<-cumsum(nn)
    re<-rs-nn+1
    n<-min(nn)
    if (!is.null(names(pvalues))) {
      grp=factor(rep(names(pvalues), nn), levels=names(pvalues))
      names(pvalues)<-NULL
    } else {
      grp=factor(rep(1:length(pvalues), nn))
    }
    pvo<-pvalues
    pvalues<-numeric(sum(nn))
    exp.x<-numeric(sum(nn))
    for(i in 1:length(pvo)) {
      if (!already.transformed) {
        pvalues[rs[i]:re[i]] <- -log10(pvo[[i]])
        exp.x[rs[i]:re[i]] <- -log10((rank(pvo[[i]], ties.method="first")-.5)/nn[i])
      } else {
        pvalues[rs[i]:re[i]] <- pvo[[i]]
        exp.x[rs[i]:re[i]] <- -log10((nn[i]+1-rank(pvo[[i]], ties.method="first")-.5)/(nn[i]+1))
      }
    }
  } else {
    n <- length(pvalues)+1
    if (!already.transformed) {
      exp.x <- -log10((rank(pvalues, ties.method="first")-.5)/n)
      pvalues <- -log10(pvalues)
    } else {
      exp.x <- -log10((n-rank(pvalues, ties.method="first")-.5)/n)
    }
  }
  
  
  #this is a helper function to draw the confidence interval
  panel.qqconf<-function(n, conf.points=1000, conf.col="gray", conf.alpha=.05, ...) {
    require(grid)
    conf.points = min(conf.points, n-1);
    mpts<-matrix(nrow=conf.points*2, ncol=2)
    for(i in seq(from=1, to=conf.points)) {
      mpts[i,1]<- -log10((i-.5)/n)
      mpts[i,2]<- -log10(qbeta(1-conf.alpha/2, i, n-i))
      mpts[conf.points*2+1-i,1]<- -log10((i-.5)/n)
      mpts[conf.points*2+1-i,2]<- -log10(qbeta(conf.alpha/2, i, n-i))
    }
    grid.polygon(x=mpts[,1],y=mpts[,2], gp=gpar(fill=conf.col, lty=0), default.units="native")
  }
  
  #reduce number of points to plot
  if (should.thin==T) {
    if (!is.null(grp)) {
      thin <- unique(data.frame(pvalues = round(pvalues, thin.obs.places),
                                exp.x = round(exp.x, thin.exp.places),
                                grp=grp))
      grp = thin$grp
    } else {
      thin <- unique(data.frame(pvalues = round(pvalues, thin.obs.places),
                                exp.x = round(exp.x, thin.exp.places)))
    }
    pvalues <- thin$pvalues
    exp.x <- thin$exp.x
  }
  gc()
  
  prepanel.qqunif= function(x,y,...) {
    A = list()
    A$xlim = range(x, y)*1.02
    A$xlim[1]=0
    A$ylim = A$xlim
    return(A)
  }
  
  #draw the plot
  xyplot(pvalues~exp.x, groups=grp, xlab=xlab, ylab=ylab, aspect=aspect,
         prepanel=prepanel, scales=list(axs="i"), pch=pch,
         panel = function(x, y, ...) {
           if (draw.conf) {
             panel.qqconf(n, conf.points=conf.points, 
                          conf.col=conf.col, conf.alpha=conf.alpha)
           };
           panel.xyplot(x,y, ...);
           panel.abline(0,1);
         }, par.settings=par.settings, ...
  )
}

qqunif.plot(pval1) #these are the raw p-values, not log-transformed



######################################
########### MANHATTAN PLOT ###########
# Based on https://danielroelfs.com/blog/how-i-create-manhattan-plots-using-ggplot/
######################################
library(tidyverse)

## read positions 
snp <- unite(p,"SNP", V1:V2 , sep = "_")
manh <- cbind(snp, p, pval1)
names(manh)<-c("SNP","CHR", "BP", "P1")
manh$BP <- as.numeric(manh$BP)
chrOrder <-c((1:25))
manh$CHR <- factor(manh$CHR, levels=chrOrder, ordered = TRUE)
manh <- manh[order(manh$CHR),]
manh_p1 <- subset(manh, select = c("SNP","CHR", "BP", "P1"))
manh_p2 <- subset(manh, select = c("SNP","CHR", "BP", "P2"))
names(manh_p1) <-c("SNP","CHR", "BP", "P")
names(manh_p2) <-c("SNP","CHR", "BP", "P")

#selecting only signicifant SNPs
manh_p1.sig.dat <- manh_p1 %>% subset(P < 0.05)

#Which chromosomes
unique(manh_p1.sig.dat$CHR)
#Then, loop through the chromosomes and add to each base pair position the latest position 
#from the previous chromosome. This will create a column in which the relative base pair 
#position is the position as if it was stitched together. For each chromosome, I extract 
#the largest base pair position, put it in a list, and then in a temporary variable, I add
#the length of the previous chromosomes together and add them to the relative base pair 
#position in the current chromosome and save it in a column called BPcum
nCHR <- length(unique(manh_p1.sig.dat$CHR))
manh_p1.sig.dat$BPcum <- NA
s <- 0
nbp <- c()
for (i in unique(manh_p1.sig.dat$CHR)){
  nbp[i] <- max(manh_p1.sig.dat[manh_p1.sig.dat$CHR == i,]$BP)
  manh_p1.sig.dat[manh_p1.sig.dat$CHR == i,"BPcum"] <- manh_p1.sig.dat[manh_p1.sig.dat$CHR == i,"BP"] + s
  s <- s + nbp[i]
}

#Center the label at each chromosome center, give y-axis enough space above and
#set Bonferroni 0.05/#ofSNPs significant level
axis.set <- manh_p1.sig.dat %>% 
  group_by(CHR) %>% 
  summarize(center = (max(BPcum) + min(BPcum)) / 2)
ylim <- abs(floor(log10(min(manh_p1.sig.dat$P)))) + 2 
#BonferroniCorr
sig <- 0.05 / nrow(manh_p1)

### PLOT ###
manhplot <- ggplot(manh_p1.sig.dat, aes(x = BPcum, y = -log10(P), 
                                 color = as.factor(CHR), size = -log10(P))) +
  geom_point(alpha = 0.75) +
  geom_hline(yintercept = -log10(sig), color = "grey40", linetype = "dashed") + 
  scale_x_continuous(label = axis.set$CHR, breaks = axis.set$center) +
  scale_y_continuous(expand = c(0,0), limits = c(2, ylim)) +
  scale_color_manual(values = rep(c("#276FBF", "#183059"), nCHR)) +
  scale_size_continuous(range = c(0.5,3)) +
  labs(x = "Chromosome", 
       y = "-log10(p)") + 
  theme_minimal() +
  theme( 
    legend.position = "none",
    panel.border = element_blank(),
    panel.grid.major.x = element_blank(),
    panel.grid.minor.x = element_blank(),
    axis.text.x = element_text(angle = 0, size = 8, vjust = 0.5)
  )

ggsave("NewManhattan_PC1_signifSnps.pdf", plot(manhplot), device = "pdf", width = 297, height = 100, units = c("mm"))

## see the position of the most significant SNP
#p[which.max(s1),]
#p[pval1<0.00000001,]
manh_p1_ordered <- manh_p1[order(manh_p1$P),]
manh_p1_ordered_top20 <- manh_p1_ordered[1:20,]
write.table(manh_p1_ordered_top20, file = "ModernPuffin.NoThu.NoZNoUnpl.pruned.pcangsd.selection.PC1.significant.top20.txt")

#use annotate() to add text to SNPs
```