# NJTree with Outgroup (RAZ)


NJ Tree - with bootstrapped p-distance matrices based on genotype likelihoods

```bash

sbatch ngsDist.sh
```


-> ngsDist.sh

```bash
#!/bin/bash

module load ngsDist/191118-GCC-8.2.0-2.31.1

NS=$(zcat ModernPuffinAngsd_NoZ_NoUnpl.beagle.ldpruned.withOut.gz | tail -n +2 | wc -l)
echo "Total No. of Sites = "$NS

ngsDist --verbose 1 --geno ModernPuffinAngsd_NoZ_NoUnpl.beagle.ldpruned.withOut.gz --n_ind 72 --n_sites $NS --probs --seed 2705 --n_threads 16 --n_boot_rep 100 --boot_block_size 20 --out ModernPuffinAngsd..NoZ_NoUnpl.WithOut.NU.100boot.dist

cp ModernPuffinAngsd..NoZ_NoUnpl.WithOut.NU.100boot.dist ../

cd ../

rm -r temp/
```

then FastME and IQtree

```bash
cd /Users/oliverke/Desktop/PhD_Bioinformatics/Mapping/nuclear/ANGSD/SAGA//NJTree/Manuscript

scp oliverke@saga.sigma2.no:/cluster/work/users/oliverke/ANGSD_Modern//GenoLikeResult/NJTree/WithOutgroup/ModernPuffinAngsd..NoZ_NoUnpl.WithOut.NU.100boot.dist .

/Users/oliverke/BioinformaticsProgs/fastme-2.1.5/binaries/fastme-2.1.5-osx -D 101 -i ModernPuffinAngsd..NoZ_NoUnpl.WithOut.NU.100boot.dist -o ModernPuffinAngsd..NoZ_NoUnpl.WithOut.NU.100boot.tree -B -m B -s -n

scp ModernPuffinAngsd..NoZ_NoUnpl.WithOut.NU.100boot.tree oliverke@saga.sigma2.no:/cluster/work/users/oliverke/ANGSD_Modern//GenoLikeResult/NJTree/WithOutgroup/

head -1 ModernPuffinAngsd..NoZ_NoUnpl.WithOut.NU.100boot.tree > ModernPuffinAngsd..NoZ_NoUnpl.WithOut.NU.100boot.main.tree
tail -n +2 ModernPuffinAngsd..NoZ_NoUnpl.WithOut.NU.100boot.tree | awk 'NF' > ModernPuffinAngsd..NoZ_NoUnpl.WithOut.NU.100boot.boot.trees

# manually root trees
sed 's/^/(/g' ModernPuffinAngsd..NoZ_NoUnpl.WithOut.NU.100boot.main.tree | sed 's/,Ind_71/):0,Ind_71/g' > ModernPuffinAngsd..NoZ_NoUnpl.WithOut.NU.100boot.main.rooted.tree
sed 's/^/(/g' ModernPuffinAngsd..NoZ_NoUnpl.WithOut.NU.100boot.boot.trees | sed 's/,Ind_71/):0,Ind_71/g' > ModernPuffinAngsd..NoZ_NoUnpl.WithOut.NU.100boot.boot.rooted.trees

#assign bootstrap
module load IQ-TREE/1.6.12-foss-2018b
iqtree -o Ind_71 -sup ModernPuffinAngsd..NoZ_NoUnpl.WithOut.NU.100boot.main.rooted.tree ModernPuffinAngsd..NoZ_NoUnpl.WithOut.NU.100boot.boot.rooted.trees
scp oliverke@saga.sigma2.no:/cluster/work/users/oliverke/ANGSD_Modern//GenoLikeResult/NJTree/WithOutgroup/ModernPuffinAngsd..NoZ_NoUnpl.WithOut.NU.100boot.boot.rooted.trees.suptree .

#create ind to ID file on local #DONE ONLY ONCE
cp ../../PCAngsd/bams_good .
awk -F '.' '{print $1}' bams_good > Tree.IDs.txt
#open file in excel and insert column before 1st column with Ind_0 to 71 + RAZ RAZ
awk '{print $1"\t"$2}' Tree.IDs.txt > Tip.IDs.txt

# Use Rscript TreeInR.R

```

#MANUSCRIPT

```bash
cd /cluster/work/users/oliverke/ANGSD_Modern//GenoLikeResult/NJTree/WithOutgroup/Manuscript

cp /cluster/work/users/oliverke/ANGSD_Modern//Assembly_Puffin_NU.MT.* .

mkdir GenoLikeResult

#create files for main script and to get region file
echo -e "Scaffolds_chromosome_10:\nScaffolds_chromosome_11:\nScaffolds_chromosome_12:\nScaffolds_chromosome_13:\nScaffolds_chromosome_14:\nScaffolds_chromosome_15:\nScaffolds_chromosome_16:\nScaffolds_chromosome_17:\nScaffolds_chromosome_18:\nScaffolds_chromosome_19:\nScaffolds_chromosome_1:\nScaffolds_chromosome_20:\nScaffolds_chromosome_21:\nScaffolds_chromosome_22:\nScaffolds_chromosome_23:\nScaffolds_chromosome_24:\nScaffolds_chromosome_25:\nScaffolds_chromosome_2:\nScaffolds_chromosome_3:\nScaffolds_chromosome_4:\nScaffolds_chromosome_5:\nScaffolds_chromosome_6:\nScaffolds_chromosome_7:\nScaffolds_chromosome_8:\nScaffolds_chromosome_9:" > ChromosomeList.NoUn.NoZ

cp /cluster/work/users/oliverke/ANGSD_Modern//GenoLikeResult/PCAngsd/ModernPuffinAngsd_NoZ_NoUnpl.beagle.ldpruned.gz .
zcat ModernPuffinAngsd_NoZ_NoUnpl.beagle.ldpruned.gz | tail -n +2 | awk '{print $1}' - | awk -F '_' '{print $1"_"$2"_"$3"\t"$4}' > positions
rm ModernPuffinAngsd_NoZ_NoUnpl.beagle.ldpruned.gz

cat ChromosomeList.NoUn.NoZ | while read line ;
do line_edit=$(echo ${line} | awk -F ':' '{print $1}') ;
grep -w ${line_edit} positions > ${line_edit}.positions ;
done

mkdir test
cd test
ls ../../../../../*.bam | grep -v "IOM001" | grep -v "THU" > bams_good_edit
cp bams_good_edit ../
cd ../
rm -r test/

#run Puffin_ANGSD_main_Outgroup.sh
cat ChromosomeList.NoUn.NoZ | grep -v "Scaffolds_chromosome_1:" | grep -v "Scaffolds_chromosome_2:" | grep -v "Scaffolds_chromosome_3:" | while read i ;
do echo ${i} ;
sbatch Puffin_ANGSD_main_Outgroup.sh ${i} ;
done

#run an alternate with more time (24hr) for chr1,2,3
for i in Scaffolds_chromosome_1: Scaffolds_chromosome_2: Scaffolds_chromosome_3: ;
do echo ${i} ;
sbatch Puffin_ANGSD_main_Outgroup.sh ${i} ;
done
```

nano Puffin_ANGSD_main_Outgroup.sh 

```bash
#!/bin/bash

#SBATCH --job-name=ANGSDOut
#SBATCH --account=nn9244k
#SBATCH --time=12:00:00
#SBATCH --cpus-per-task=16
#SBATCH --mem-per-cpu=5000M
#SBATCH --mail-type=ALL
#SBATCH --mail-user=okerstenwork@gmail.com

#if you need long time
# --partition=long
#If you need hugemem
# --partition=hugemem

##Submission Variables
date
pwd -L
module purge
set -o errexit

Chromosome=$(echo ${1} | awk -F ':' '{print $1}' | awk -F '_' '{print $2"_"$3}')
ChromosomeAlt=$(echo ${1} | awk -F ':' '{print $1}')

echo ${Chromosome}
echo ${ChromosomeAlt}

mkdir temp.${Chromosome}

cp bams_good_edit temp.${Chromosome}/
cp ${ChromosomeAlt}.positions temp.${Chromosome}/
cp Assembly_Puffin_NU.MT* temp.${Chromosome}/

cd temp.${Chromosome}/

FILTERS="-uniqueOnly 1 -remove_bads 1 -minMapQ 30 -minQ 30 -dosnpstat 1 -C 50 -baq 2 -checkBamHeaders 1 -doHWE 1 -skipTriallelic 1"
#only add filters against trialleles. The remaining SNPs should be fine as filtered before

# -doMajorMinor 5: force the major allelel according to your ancestral states if you have defined those -anc. The minor allele will be inferred based on the genotype likelihood (see do major minor 1)
TODO="-doMajorMinor 1 -doMaf 1 -doCounts 1 -doGeno 8 -doPost 1 -doGlf 2"

# Run command:
module load angsd/0.931-GCC-8.2.0-2.31.1

Threads="16"

#index the sites file
echo ${1}

sleep 120

angsd sites index ${ChromosomeAlt}.positions

sleep 120

angsd -b bams_good_edit -GL 1 $FILTERS $TODO -P ${Threads} -ref Assembly_Puffin_NU.MT.fasta -r ${1} -sites ${ChromosomeAlt}.positions -out ModernPuffinAngsd_Outgroup_${Chromosome}

mv ModernPuffinAngsd_Outgroup_${Chromosome}* ../GenoLikeResult/

cd ../

rm -r temp.${Chromosome}/
```

```bash
#combine them all
cd /cluster/work/users/oliverke/ANGSD_Modern//GenoLikeResult/NJTree/WithOutgroup/Manuscript/GenoLikeResult
#No Z or unplaced
zcat ModernPuffinAngsd_Outgroup_chromosome_1.beagle.gz > ModernPuffinAngsd_Outgroup_NoZ_NoUn.beagle 
for i in 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 2 3 4 5 6 7 8 9 ;
do zcat ModernPuffinAngsd_Outgroup_chromosome_${i}.beagle.gz | tail -n +2 >> ModernPuffinAngsd_Outgroup_NoZ_NoUn.beagle ;
done ;
gzip ModernPuffinAngsd_Outgroup_NoZ_NoUn.beagle
zcat ModernPuffinAngsd_Outgroup_NoZ_NoUn.beagle.gz | tail -n +2 | wc -l
#->1,093,765 SNPs
#->RAZ not covered on 845,293 (77.28%) SNPs

mv ModernPuffinAngsd_Outgroup_NoZ_NoUn.beagle.gz ../

cd ../
rm -r GenoLikeResult/

sbatch ngsDist.sh

```

##########NJ Tree - with bootstrapped p-distance matrices based on genotype likelihoods########

nano ngsDist.sh

```bash
#!/bin/bash

#SBATCH --job-name=ngsDist
#SBATCH --account=nn9244k
#SBATCH --time=05:00:00
#SBATCH --cpus-per-task=16
#SBATCH --mem-per-cpu=7000M

#if you need long time
# --partition=long
#If you need hugemem
# --partition=hugemem

##Submission Variables
date
pwd -L
module purge
set -o errexit

mkdir temp
cp ModernPuffinAngsd_Outgroup_NoZ_NoUn.beagle.gz temp/

cd temp/

module load ngsDist/191118-GCC-8.2.0-2.31.1

NS=$(zcat ModernPuffinAngsd_Outgroup_NoZ_NoUn.beagle.gz | tail -n +2 | wc -l)
echo "Total No. of Sites = "$NS

ngsDist --verbose 1 --geno ModernPuffinAngsd_Outgroup_NoZ_NoUn.beagle.gz --n_ind 72 --n_sites $NS --probs --seed 2705 --n_threads 16 --n_boot_rep 100 --boot_block_size 20 --out ModernPuffinAngsd..NoZ_NoUnpl.WithOut.NU.100boot.dist

cp ModernPuffinAngsd..NoZ_NoUnpl.WithOut.NU.100boot.dist ../

cd ../

rm -r temp/
```

then FastME and IQtree

```bash
cd /Users/oliverke/Desktop/PhD_Bioinformatics/Mapping/nuclear/ANGSD/SAGA//NJTree/Manuscript/Better

scp oliverke@saga.sigma2.no:/cluster/work/users/oliverke/ANGSD_Modern//GenoLikeResult/NJTree/WithOutgroup/Manuscript/ModernPuffinAngsd..NoZ_NoUnpl.WithOut.NU.100boot.dist .

/Users/oliverke/BioinformaticsProgs/fastme-2.1.5/binaries/fastme-2.1.5-osx -D 101 -i ModernPuffinAngsd..NoZ_NoUnpl.WithOut.NU.100boot.dist -o ModernPuffinAngsd..NoZ_NoUnpl.WithOut.NU.100boot.tree -B -m B -s -n

cat ModernPuffinAngsd..NoZ_NoUnpl.WithOut.NU.100boot.tree | awk 'NF' > ModernPuffinAngsd..NoZ_NoUnpl.WithOut.NU.100boot.edit.tree 

#Reroot all trees 
touch ModernPuffinAngsd..NoZ_NoUnpl.WithOut.NU.100boot.rooted.tree 
for i in {1..101} ; 
do echo ${i} ;
head -${i} ModernPuffinAngsd..NoZ_NoUnpl.WithOut.NU.100boot.edit.tree | tail -1 > ModernPuffinAngsd..NoZ_NoUnpl.WithOut.NU.100boot.edit.tree.${i} ;
Rscript Reroot.R ModernPuffinAngsd..NoZ_NoUnpl.WithOut.NU.100boot.edit.tree.${i} ;
rm ModernPuffinAngsd..NoZ_NoUnpl.WithOut.NU.100boot.edit.tree.${i} ;
cat Test >> ModernPuffinAngsd..NoZ_NoUnpl.WithOut.NU.100boot.rooted.tree ;
rm Test ;
done 

scp ModernPuffinAngsd..NoZ_NoUnpl.WithOut.NU.100boot.rooted.tree oliverke@saga.sigma2.no:/cluster/work/users/oliverke/ANGSD_Modern//GenoLikeResult/NJTree/WithOutgroup/Manuscript
head -1 ModernPuffinAngsd..NoZ_NoUnpl.WithOut.NU.100boot.rooted.tree > ModernPuffinAngsd..NoZ_NoUnpl.WithOut.NU.100boot.main.rooted.tree
tail -n +2 ModernPuffinAngsd..NoZ_NoUnpl.WithOut.NU.100boot.rooted.tree | awk 'NF' > ModernPuffinAngsd..NoZ_NoUnpl.WithOut.NU.100boot.boot.rooted.trees
#assign bootstrap
module load IQ-TREE/1.6.12-foss-2018b
iqtree -o Ind_53 -sup ModernPuffinAngsd..NoZ_NoUnpl.WithOut.NU.100boot.main.rooted.tree ModernPuffinAngsd..NoZ_NoUnpl.WithOut.NU.100boot.boot.rooted.trees
scp oliverke@saga.sigma2.no:/cluster/work/users/oliverke/ANGSD_Modern//GenoLikeResult/NJTree/WithOutgroup/Manuscript/ModernPuffinAngsd..NoZ_NoUnpl.WithOut.NU.100boot.boot.rooted.trees.suptree .

#create ind to ID file on local #DONE ONLY ONCE
cp ../../PCAngsd/bams_good .
awk -F '.' '{print $1}' bams_good > Tree.IDs.txt
#open file in excel and insert column before 1st column with Ind_0 to 71 + RAZ RAZ
awk '{print $1"\t"$2}' Tree.IDs.txt > Tip.IDs.txt

# Use Rscript TreeInR.R

```