# SFS + nucDiv/TajimaD + Fst 

WORKFLOW

```bash
1. Create sites to analyze by running a normal ANGSD run without filters 
that distort allele frequency

2. Use those sites to run a -doSAF ANGSD run to get a "saf.idx" file for each 
population AND globally AND genomic cluster

3. Use realSFS to convert the "saf.idx" file to a ".sfs" file
	- globally
	- for each population
	- for each genomic cluster

### 1D SFS ###
#Theta + Tajima
1. Use the ".sfs" as 'pest' input for a ANGSD -doSAF -doThetas run
2. Use thetaStat do_stat to get a "per chromosome" calculation of varius statistics
-> The output of 1. are log scaled per site estimates of various thetas
-> the output of 2. are the sum of the per site estimates for each chromosome

### 2D SFS ###
#Fst
1. Get prior for two pops using realSFS and the two "saf.idx"
2. Get the global Fst with that prior and the previous "saf.idx" with realSFS 
index fst (and method -which Fst) 
```

Let's GO

```bash
grep -v "Z" ChromosomeList | grep -v "unplaced" > ChromosomeList.edit

# make neccessary pop files
ls ../../*.bam | grep -v "IOM001" | grep -v "THU" | grep -v "RAZ" > bams_good
cp bams_good Global.bamlist

#running a global ANGSD run for all sites to get sites info
cat ChromosomeList.edit | while read i ;
do echo ${i} ;
sbatch Puffin_ANGSD1_SFS.sh ${i} ;
done

```

-> Puffin_ANGSD1_SFS.sh

```bash
#!/bin/bash

Chromosome=$(echo ${1} | awk -F ':' '{print $1}' | awk -F '_' '{print $2"_"$3}')

FILTERS="-uniqueOnly 1 -remove_bads 1 -minMapQ 30 -minQ 30 -dosnpstat 1 -C 50 -baq 2 -checkBamHeaders 1 -doHWE 1 -sb_pval 1e-5 -hetbias_pval 1e-5 -skipTriallelic 1 -minInd 71 -setmaxDepth 635 -setminDepth 365"
TODO="-doMajorMinor 1 -doMaf 1 -doCounts 1 -doGeno 3 -doPost 2"


module load angsd/0.931-GCC-8.2.0-2.31.1

Threads="16"

angsd -b Global.bamlist -GL 1 $FILTERS $TODO -P ${Threads} -ref Assembly_Puffin_NU.MT.fasta -r ${1} -out ModernPuffinAngsd_SFS_${Chromosome}

```

Merging all sites

```bash

zcat ModernPuffinAngsd_SFS_chromosome_1.snpStat.gz > sfsSites.snpStat
for i in 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 2 3 4 5 6 7 8 9 ;
do zcat ModernPuffinAngsd_SFS_chromosome_${i}.snpStat.gz | tail -n +2 >> sfsSites.snpStat ;
done ;

```

Removing "paralogs"
```bash
zcat sfsSites.snpStat.gz | awk '($3+$4+$5+$6)>0' | awk '($12+$13+$14+$15)/($3+$4+$5+$6)<0.5' | cut -f 1,2 > sites2do
```

Estimate folded saf.idx from sites2do (globally and pop-wise, cluster-wise)

```bash

# make neccessary pop files
ls ../../../*.bam | grep -v "IOM001" | grep -v "THU" | grep -v "RAZ" > bams_good
cp bams_good Global.bamlist
#colonies
for i in BJO BRE FAR GRI HOR IOM PAP ROS SPI WES GAN GUL ;
do grep ${i} bams_good > ${i}.bamlist ;
done
# genomic clusters, BJO, SPI, IOM, CAN, MAI
touch MAI.bamlist
for i in BRE FAR GRI HOR PAP ROS WES ;
do grep ${i} bams_good >> MAI.bamlist ;
done ;
touch CAN.bamlist
for i in GAN GUL ;
do grep ${i} bams_good >> CAN.bamlist ;
done ;

ls *.bamlist | while read line ;
do echo ${line} ;
sbatch SFS_Angsd.sh ${line} ;
done

```

-> SFS_Angsd.sh

```bash
#!/bin/bash

module purge 
module load SAMtools/1.9-GCC-8.2.0-2.31.1
samtools faidx Assembly_Puffin_NU.MT.fasta

module load angsd/0.931-GCC-8.2.0-2.31.1

Threads="1"
GENOME_ANC=Assembly_Puffin_NU.MT.fasta
GENOME_REF=Assembly_Puffin_NU.MT.fasta
Population=$(echo ${1} | awk -F '.' '{print $1}')

TODO="-doSaf 1"

Chromosome=$(cat ChromosomeList.edit | awk -F ':' '{print $1}' | grep -w Scaffolds_chromosome_${SLURM_ARRAY_TASK_ID} )
ChromosomeEdit=$(echo ${Chromosome}":")
grep -w Scaffolds_chromosome_${SLURM_ARRAY_TASK_ID} ../sites2do > sites_Chromosome${SLURM_ARRAY_TASK_ID}_2do

sleep 120

angsd sites index sites_Chromosome${SLURM_ARRAY_TASK_ID}_2do

sleep 120 

angsd -r ${ChromosomeEdit} -sites sites_Chromosome${SLURM_ARRAY_TASK_ID}_2do -b ${1} -GL 1 -P ${Threads} -anc ${GENOME_ANC} -ref ${GENOME_REF} $TODO -out ${Population}_Chromosome${SLURM_ARRAY_TASK_ID}

```
Estimate folded SFS from saf.idx (globally and pop-wise, cluster-wise)


```bash

for i in BJO BRE FAR GRI HOR IOM PAP ROS SPI WES GAN GUL CAN Global MAI ;
do sbatch SFS_Calc.sh ${i};
done
```

-> SFS_Calc.sh

```bash
#!/bin/bash
module load angsd/0.931-GCC-8.2.0-2.31.1
Threads="16"
Population=$(echo ${1})

realSFS cat ${Population}_Chromosome*.saf.idx -outnames ${Population} 
realSFS ${Population}.saf.idx -P ${Threads} -fold 1 > ${Population}.sfs 

```

NucDiv/TajimasD calculation

```bash

for i in BJO BRE FAR GRI HOR IOM PAP ROS SPI WES GAN GUL Global CAN MAI ;
do sbatch SFS_Theta.sh ${i} ;
done

```

-> SFS_Theta.sh 

```bash
#!/bin/bash

Population=$(echo ${1})

mkdir temp.${Population}

module purge
module load angsd/0.931-GCC-8.2.0-2.31.1
Threads="16"

realSFS saf2theta ${Population}.saf.idx -outname ${Population} -sfs ${Population}.sfs -fold 1 -P ${Threads}

#calculate Tajimas D per chromosome
thetaStat do_stat ${Population}.thetas.idx

```

Fst

```bash

POPS="BJO BRE FAR GRI HOR IOM PAP ROS SPI WES GAN GUL"

LIST=$POPS
dequeue_from_list() {
  shift;
  LIST=$@
}
for pop1 in $POPS ; 
do dequeue_from_list $LIST ;
for pop2 in $LIST ; 
do echo ${pop1}" "${pop2} ;
sbatch SFS_Fst.sh ${pop1} ${pop2} ;
done ;
done ;

for pop1 in IOM SPI;
do for pop2 in CAN MAI ; # 24hr with MAI
do echo ${pop1}" "${pop2} ;
sbatch SFS_Fst.sh ${pop1} ${pop2};
done ;
done ;

sbatch SFS_Fst.sh CAN MAI;

```

-> SFS_Fst.sh

```bash
#!/bin/bash

module purge
module load angsd/0.931-GCC-8.2.0-2.31.1
Threads="1"


realSFS ${1}.saf.idx ${2}.saf.idx -P ${Threads} -fold 1 > ${1}.${2}.ml
realSFS fst index ${1}.saf.idx ${2}.saf.idx -sfs ${1}.${2}.ml -whichFst 1 -fstout ${1}.${2} -P ${Threads} -fold 1
realSFS fst stats ${1}.${2}.fst.idx -P ${Threads} -fold 1 > ${1}.${2}.Fst

```

Combine Results

```bash

touch Colony.Fst.txt
POPS="BJO BRE FAR GRI HOR IOM PAP ROS SPI WES GAN GUL"

LIST=$POPS
dequeue_from_list() {
  shift;
  LIST=$@
}
for pop1 in $POPS ; 
do dequeue_from_list $LIST ;
for pop2 in $LIST ; 
do echo ${pop1}" "${pop2} ;
Fst=$(awk '{print $2}' ${pop1}.${pop2}.Fst)
echo -e ${pop1}"\t"${pop2}"\t"${Fst} >> Colony.Fst.txt
done ;
done ; 

touch Cluster.Fst.txt
for pop1 in IOM SPI ;
do for pop2 in CAN MAI ;
do echo ${pop1}" "${pop2} ;
Fst=$(awk '{print $2}' ${pop1}.${pop2}.Fst)
echo -e ${pop1}"\t"${pop2}"\t"${Fst} >> Cluster.Fst.txt
done ;
done ; 
pop1=CAN
pop2=MAI
Fst=$(awk '{print $2}' ${pop1}.${pop2}.Fst)
echo -e ${pop1}"\t"${pop2}"\t"${Fst} >> Cluster.Fst.txt

```