# SNP annotation

#ON LOCAL

```bash
# building files for snpEff run
# edit the config file to add genomes and codon table

# nano snpEff.config

# in snpEff data folder, add
# mkdir genomes
# cd genomes
# cp Assembly_Puffin_MT.final.masked.fasta .
# mv Assembly_Puffin_MT.final.masked.fasta Puffin_mt_1.0.fa
# cd ..
# mkdir Puffin_mt_1.0
# cd Puffin_mt_1.0

# make a copy of the gff file (scaffold1766-2.gff) and add header,
# CDS for each gene with ID + parent, ncRNA for each RNA + ID and parent (see example gff)
# also change gene extents - extend ND1+4, change middle of ND3, remove T from COX3 end

# cp scaffold1766-2.gff .
# mv scaffold1766-2.gff > genes.gff

# cd snpEff
# java -jar snpEff.jar build -gff3 -v Puffin_mt_1.0
# java -jar snpEff.jar dump Puffin_mt_1.0 | less

# run snpEff
cd SNPAnnotation/2020_03_05

java -Xmx4g -jar snpEff.jar -c snpEff.config -ud 0 Puffin_mt_1.0 PuffinMT_noOut_SNPs.filtered.vcf > PuffinMT_noOut_SNPs.filtered.annot.vcf

#####STATS#######
echo "# of SNPs"
grep -v "#" PuffinMT_noOut_SNPs.filtered.annot.vcf | wc -l
echo "# of Intergenic SNPs"
grep -v "#" PuffinMT_noOut_SNPs.filtered.annot.vcf | grep "intergenic_region" - | wc -l
echo "# SNPs in gene regions"
grep -v "#" PuffinMT_noOut_SNPs.filtered.annot.vcf | grep -v "intergenic_region" - | wc -l
echo "# SNPs in non-coding regions (tRNA or rRNA)"
grep -v "#" PuffinMT_noOut_SNPs.filtered.annot.vcf | grep "non_coding_transcript_variant" - | wc -l
echo "# of SNPs in coding regions (CDS)"
grep -v "#" PuffinMT_noOut_SNPs.filtered.annot.vcf | grep -v "intergenic_region" - | grep -v "non_coding_transcript_variant" - | wc -l
echo "# of synonymous SNPs"
grep -v "#" PuffinMT_noOut_SNPs.filtered.annot.vcf | grep "synonymous" - | wc -l
echo "# of non-synonymous SNPs"
grep -v "#" PuffinMT_noOut_SNPs.filtered.annot.vcf | grep -v "intergenic_region" - | grep -v "non_coding_transcript_variant" - | grep -v "synonymous" - | wc -l
echo "# of missense SNPs"
grep -v "#" PuffinMT_noOut_SNPs.filtered.annot.vcf | grep "missense" - | wc -l
echo "# of nonsense SNPs"
grep -v "#" PuffinMT_noOut_SNPs.filtered.annot.vcf | grep "stop_gained" - | wc -l
echo "# of splice regions"
grep -v "#" PuffinMT_noOut_SNPs.filtered.annot.vcf | grep "splice_region_variant" - | wc -l

##BETTER VCF TABLE
grep -v "#" PuffinMT_noOut_SNPs.filtered.annot.vcf | awk -F 'AC=' '{print $1}' > part1
grep -v "#" PuffinMT_noOut_SNPs.filtered.annot.vcf | awk -F 'ANN=' '{print $2}' > part2
echo -e "CHROM\tPOS\tREF\tALT\tQUAL\tFILTER\tAllele\tAnnotation\tAnnotation_Impact\tGene_Name\tNucleotideChange\tAminoAcidChange\tNucleotidePosition\tAminoAcidPosition" > PuffinMT_noOut_SNPs.filtered.annot.table.txt
paste part1 part2 | awk '{print $1"\t"$2"\t"$4"\t"$5"\t"$6"\t"$7"\t"$8}' - | awk -F '|' '{print $1"\t"$2"\t"$3"\t"$4"\t"$10"\t"$11"\t"$12"\t"$14}' >> PuffinMT_noOut_SNPs.filtered.annot.table.txt
rm part*
```