# VCF filter

Prepping Raw VCF

```bash

#Total Variants (without Outgroup)
gunzip NoThulePuffin_MT_SnpsAndIndels.raw.vcf.gz
module purge
module load VCFtools/0.1.16-intel-2018b-Perl-5.28.0
module load BCFtools/1.9-intel-2018b
vcftools --vcf NoThulePuffin_MT_SnpsAndIndels.raw.vcf --remove-indels --out NoThuPuffin_MT_SnpsAndIndels.raw

```

Analysis of Puffin VCF

```bash
# filtering
bcftools filter --SnpGap 10 -e 'QD < 2.0 || MQ < 40 || FS > 60.0 || SOR > 3 || MQRankSum < -12.5 || ReadPosRankSum < -8.0' -o NoThu_MT_SnpsAndIndels.filtered1.vcf NoThulePuffin_MT_SnpsAndIndels.raw.vcf
# More filtering on Depth and genotypeQuality
bcftools filter -S . -e 'FMT/DP<3 | FMT/GQ<15' -o NoThu_MT_SnpsAndIndels.filtered2.vcf NoThu_MT_SnpsAndIndels.filtered1.vcf
# Filtering on monomorphic SNPs and remove indels - only keep biallelic
bcftools filter -e 'AC==0 || AC==AN' NoThu_MT_SnpsAndIndels.filtered2.vcf | bcftools view -m2 -M2 -v snps -o NoThu_MT_Snps.filtered1.vcf
bcftools view -H NoThu_MT_Snps.filtered1.vcf | wc -l


# check samples with missing genotypes and sites with many missing genotypes -> remove unusual samples(25%)/sites(20%, 0.8)
vcftools --vcf NoThu_MT_Snps.filtered1.vcf --missing-indv --out NoThu_MT_Filt_SNPs_stats
cat NoThu_MT_Filt_SNPs_stats.imiss | sort -rk5 | head


vcftools --vcf NoThu_MT_Snps.filtered1.vcf --missing-site --out NoThu_MT_Filt_SNPs_stats
cat NoThu_MT_Filt_SNPs_stats.lmiss | sort -rk6 | head 


# remove sites with missing data
vcftools --vcf NoThu_MT_Snps.filtered1.vcf --max-missing 1.0 --recode --recode-INFO-all --out NoThu_MT_Snps.filtered2.vcf
mv NoThu_MT_Snps.filtered2.vcf.recode.vcf NoThu_MT_Snps.filtered2.vcf
# if nothing to remove, do:
# cp NoThu_MT_Snps.filtered1.vcf NoThu_MT_Snps.filtered2.vcf


# no. alt allele in only 1 sample (singletons)
vcftools --vcf NoThu_MT_Snps.filtered2.vcf --singletons --out NoThu_MT_Filt_SNPs_stats
cat NoThu_MT_Filt_SNPs_stats.singletons | tail -n+2 | grep -w "S" | wc -l

cat NoThu_MT_Filt_SNPs_stats.singletons | tail -n+2 | awk '{print $5}' | sort | uniq -c | sort -rnk1 | head


# sample wise stats
cat Modern.samples | grep -vw "RAZ" | grep -v "THU" > samples.list
bcftools stats -S samples.list NoThu_MT_Snps.filtered2.vcf > NoThu.MT.SNPs.stats.persample.txt

#Remove IOM001 and redo stats
vcftools --vcf NoThu_MT_Snps.filtered2.vcf --remove-indv IOM001 --recode --recode-INFO-all --out NoThu_MT_Snps.filtered3.vcf
# use vcflib vcffixup to correct AC, AF and NS 
export PATH=miniconda3/bin:$PATH
source conda.sh
source activate vcflib
vcffixup NoThu_MT_Snps.filtered3.vcf.recode.vcf > NoThu_MT_Snps.filtered3.edit.vcf
conda deactivate

# Exclude monomorphic SNPs from Puffins again
bcftools filter -e 'AC==0 || AC==AN' NoThu_MT_Snps.filtered3.edit.vcf | bcftools view -m2 -M2 -v snps -o NoThu_MT_Snps.filtered4.vcf
bcftools view -H NoThu_MT_Snps.filtered4.vcf | wc -l


# check samples with missing genotypes and sites with many missing genotypes -> remove unusual samples(25%)/sites(20%, 0.8)
vcftools --vcf NoThu_MT_Snps.filtered4.vcf --missing-indv --out NoThu_NoIOM_MT_Filt_SNPs_stats
cat NoThu_NoIOM_MT_Filt_SNPs_stats.imiss | sort -rk5 | head

vcftools --vcf NoThu_MT_Snps.filtered4.vcf --missing-site --out NoThu_NoIOM_MT_Filt_SNPs_stats
cat NoThu_NoIOM_MT_Filt_SNPs_stats.lmiss | sort -rk6 | head 


# remove sites with missing data
vcftools --vcf NoThu_MT_Snps.filtered4.vcf --max-missing 1.0 --recode --recode-INFO-all --out NoThu_MT_Snps.filtered5.vcf
grep -v "#" NoThu_MT_Snps.filtered5.vcf.recode.vcf | wc -l


# no. alt allele in only 1 sample (singletons)
vcftools --vcf NoThu_MT_Snps.filtered5.vcf.recode.vcf --singletons --out NoThu_NoIOM_MT_Filt_SNPs_stats
cat NoThu_NoIOM_MT_Filt_SNPs_stats.singletons | tail -n+2 | grep -w "S" | wc -l

cat NoThu_NoIOM_MT_Filt_SNPs_stats.singletons | tail -n+2 | awk '{print $5}' | sort | uniq -c | sort -rnk1 | head


# sample wise stats
cat Modern.samples | grep -vw "RAZ" | grep -v "THU" | grep -v "IOM001" > samples.edit.list
bcftools stats -S samples.edit.list NoThu_MT_Snps.filtered5.vcf.recode.vcf > NoThu.NoIOM.MT.SNPs.stats.persample.txt

# Final
cp NoThu_MT_Snps.filtered5.vcf.recode.vcf PuffinMT_noOut_SNPs.filtered.vcf

bgzip -c -i PuffinMT_noOut_SNPs.filtered.vcf > PuffinMT_noOut_SNPs.filtered.vcf.gz
bcftools index PuffinMT_noOut_SNPs.filtered.vcf.gz

cat samples.list | grep -v "RAZ" - | grep -v "THU" - | grep -v "IOM001" - | while read line ;
do sample=$(echo ${line} | awk -F '.' '{print $1}') ;
echo ${sample} ;
bcftools consensus -s ${sample} -H 1 -M N -f Assembly_Puffin_mitogenome.final.fasta PuffinMT_noOut_SNPs.filtered.vcf.gz > ${sample}.CompleteMitoRef.SNPs.fasta ;
done ;
touch CompleteMitoRef.SNPs.AllPuffin.fasta
cat samples.list | grep -v "RAZ" - | grep -v "THU" - | grep -v "IOM001" - | while read line ;
do sample=$(echo ${line} | awk -F '.' '{print $1}') ;
cat ${sample}.CompleteMitoRef.SNPs.fasta | grep -v "^>" - | awk -v VAR=${sample} 'BEGIN { ORS=""; print ">"VAR"\n" } { print $0"\n"}' >> CompleteMitoRef.SNPs.AllPuffin.fasta ;
done
rm *.CompleteMitoRef.SNPs.fasta

```