# Fitchi Haplotype Network

Run Fitchi with gene tree

Initial Prep

```bash
# fasta nexus/fasta/phylip (for Fitchi) #used from https://github.com/ODiogoSilva/ElConcatenero
module purge
module loadÂ Python/3.7.2-GCCcore-8.2.0
python ElConcatenero.py -m N -if fasta -of nexus -in All.CompleteMitoRef.SNPs.withOut.NoNs.edit.fasta -o All.CompleteMitoRef.SNPs.withOut.NoNs.edit

sed -i 's/mixed ()/DNA/g' All.CompleteMitoRef.SNPs.withOut.NoNs.edit.nex
sed -i 's/interleave=no//g' All.CompleteMitoRef.SNPs.withOut.NoNs.edit.nex


#Run RemoveBootstrap.R
library(ape) # Load the ape package
tree1 <- read.tree("All.CompleteMitoRef.SNPs.edit.nooutlier.fasta.treefile") # A normal Newick with bootstraps from RAxML, for example
tree1$node.label <- NULL # Erase the bootstrap values from the phylo object
write.tree(tree1, file = "All.CompleteMitoRef.SNPs.edit.nooutlier.nobootstrap.treefile") # Save it


# add tree at end of nexus from treefile in textwrangler
begin trees;

tree ml = [&R] >TREE<

end;
#manually change the outlier names to something short and make space uniform

#change sample names to include pops
for i in BJO BRE FAR GRI HOR IOM PAP ROS SPI WES GAN GUL ;
do sed -i '' "s/${i}/${i}_/g" All.CompleteMitoRef.SNPs.withOut.NoNs.haplotypes.edit.nex ;
done ;

source activate fitchi

python3 fitchi.py All.CompleteMitoRef.SNPs.withOut.NoNs.haplotypes.edit.nex All.SNPs.pops.withOut.fitchi.html --haploid -m 0.002 -p BJO BRE FAR GRI HOR IOM PAP ROS SPI WES GAN GUL

python3 fitchi_extract.py -e svg All.SNPs.pops.withOut.fitchi.html > All.SNPs.pops.withOut.fitchi.svg

conda deactivate

```