# Fitchi Haplotype Network


-Run Fitchi with gene tree

Initial Prep

```bash

# fasta nexus/fasta/phylip (for Fitchi) #used from https://github.com/ODiogoSilva/ElConcatenero
module purge
module loadÂ Python/3.7.2-GCCcore-8.2.0
python /cluster/home/oliverke/Programs/ElConcatenero/ElConcatenero.py -m N -if fasta -of nexus -in All.CompleteMitoRef.SNPs.withOut.NoNs.edit.fasta -o All.CompleteMitoRef.SNPs.withOut.NoNs.edit

sed -i 's/mixed ()/DNA/g' All.CompleteMitoRef.SNPs.withOut.NoNs.edit.nex
sed -i 's/interleave=no//g' All.CompleteMitoRef.SNPs.withOut.NoNs.edit.nex
```

LOCAL

Copy fasta, nexus and treefile to local

```bash


#then
cp All.CompleteMitoRef.SNPs.withOut.NoNs.edit.nex All.CompleteMitoRef.SNPs.withOut.NoNs.haplotypes.nex
# add tree at end of nexus from treefile in textwrangler
begin trees;

tree ml = [&R] >TREE<

end;
#manually change the outlier names to something short and make space uniform

#change sample names to include pops
cp All.CompleteMitoRef.SNPs.withOut.NoNs.haplotypes.nex All.CompleteMitoRef.SNPs.withOut.NoNs.haplotypes.edit.nex

for i in BJO BRE FAR GRI HOR IOM PAP ROS SPI WES GAN GUL ;
do sed -i '' "s/${i}/${i}_/g" All.CompleteMitoRef.SNPs.withOut.NoNs.haplotypes.edit.nex ;
done ;

source activate fitchi

python3 fitchi.py All.CompleteMitoRef.SNPs.withOut.NoNs.haplotypes.edit.nex All.SNPs.pops.withOut.fitchi.html --haploid -m 0.002 -p BJO BRE FAR GRI HOR IOM PAP ROS SPI WES GAN GUL

# -m For purely aesthetic reasons, you might want to increase or decrease the size of all nodes in relation to the edge lengths connecting them. This can be done by
# specifying a scale factor for all radi with the "-m" option (for example "-m 2.5").
# You could also try "-m auto", which tells Fitchi to try to find an ideal size automatically so that no two nodes are overlapping.
# -n minimum node size for display can be specified with the "-n" option
# -e specify a minimum edge length for display in the graph
# -x Another option to reduce graph complexity is to ignore all transitions and only use transversions to calculate edge lengths, which can be chosen with "-x".
# Like "-e" and "-n", this only affects the graph, not the alignment statistics.

python3 fitchi_extract.py -e svg All.SNPs.pops.withOut.fitchi.html > All.SNPs.pops.withOut.fitchi.svg

conda deactivate

```

```bash

# No. of substitutions to common ancestor of outgroup
-check which nodes, and pairwise align the sequences of those nodes in Jalview
-copy paste alignment into textwrangler and count No. of "|" 
-substract that number from alignment length, i.e. 16931
```